bon/src/lib.rs

```diff
@@ -33,11 +34,45 @@ pub mod private;
 /// This macro doesn't support `vec![expr; N]` syntax, since it's simpler to
 /// just write `vec![expr.into(); N]` using [`std::vec!`] instead.
 #[macro_export]
+#[cfg(any(test, feature = "std"))]
 macro_rules! vec {
     () => (::std::vec::Vec::new());
     ($($item:expr),+ $(,)?) => (::std::vec![$(::core::convert::Into::into($item)),+ ]);
 }
 
+/// Same as [`alloc::vec!`] but converts each element with [`Into::into()`].
+///
+/// **WARNING:** it's not recommended to import this macro into scope. Reference it
+/// using the full path (`bon::vec![]`) to avoid confusion with the [`alloc::vec!`] macro.
+///
+/// A good example of the use case for this macro is when you want to create a
+/// [`Vec<String>`] where part of the items are hardcoded string literals of type
+/// `&str` and the other part is made of dynamic [`String`] values.
+///
+/// ```
+/// fn convert_media(input_extension: &str, output_extension: &str) -> std::io::Result<()> {
+///     let ffmpeg_args: Vec<String> = bon::vec![
+///         "-i",
+///         format!("input.{input_extension}"),
+///         "-y",
+///         format!("output.{output_extension}"),
+///     ];
+///
+///     std::process::Command::new("ffmpeg").args(ffmpeg_args).output()?;
+///
+///     Ok(())
+/// }
+/// ```
+///
+/// This macro doesn't support `vec![expr; N]` syntax, since it's simpler to
+/// just write `vec![expr.into(); N]` using [`std::vec!`] instead.
+#[macro_export]
+#[cfg(all(feature = "alloc", not(any(test, feature = "std"))))]
+macro_rules! vec {
+    () => (::alloc::vec::Vec::new());
+    ($($item:expr),+ $(,)?) => (::std::vec![$(::alloc::convert::Into::into($item)),+ ]);
```

Comment:
I need to dig a bit more into this, but my immediate reaction is "why does this macro have to be duplicated"? Is there a way to avoid this? Shouldn't `std` automatically imply `alloc` is available?

Reply:
I absolutely agree that it should!

[But it is not](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=45b502af665cd5afd3dcbac59699b783)

Reply:
So the problem is that `vec/Vec` lives in `alloc` crate, which isn't accessible by default? That's really inconvenient.. I'm sure other Rust crates have some good ways of fighting with this..

I took a look at `serde`, and here is [what it does](https://github.com/serde-rs/serde/blob/e08c5de5dd62571a5687f77d99609f9d9e60784e/serde/src/lib.rs#L211-L214).

I've also been thinking of using a conditionally-compiled alias. We can just add aliases like this to `private` module (which is `pub` but `doc(hidden)`, and references them in the macro as `$crate::private::Vec/vec`

Reply:
Btw. serde also adds `extern crate alloc` under the `alloc` feature:
https://github.com/serde-rs/serde/blob/e08c5de5dd62571a5687f77d99609f9d9e60784e/serde/src/lib.rs#L161-L162

Reply:
Yes, so do we, but just for tests. The crate is not exported (which it is also not in serde).

One other way of doing things that would avoid the duplication is to `pub extern crate alloc;` (Without the `pub extern`, you cannot `pub use` it - and once it is `pub extern`, there is no need to `pub use` it.). Would you like me to push an example?

Reply:
pub extern crate works fine as well. Everything inside of the private module is considered private, eo feelf free to pub {anything} ther3

Reply:
I see you added it at the crate root. Is this the limitation of `extern crate` that it needs to be at the crate root? Anyway this way is also fine as long as that symbol is `doc(hidden)`

Reply:
Done (and moved to private)